engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/EnsoHashMap.java

```diff
@@ -0,0 +1,175 @@
+@ExportLibrary(InteropLibrary.class)
+@Builtin(stdlibName = "Standard.Base.Data.Map.Map", name = "Map")
+public final class EnsoHashMap implements TruffleObject {
+  private final EnsoHashMapBuilder mapBuilder;
+  /**
+   * Size of this Map. Basically an index into {@link EnsoHashMapBuilder}'s storage. See {@link
+   * #isEntryInThisMap(StorageEntry)}.
+   */
+  private final int snapshotSize;
+
+  private Object cachedVectorRepresentation;
+
+  private EnsoHashMap(EnsoHashMapBuilder mapBuilder, int snapshotSize) {
+    this.mapBuilder = mapBuilder;
+    this.snapshotSize = snapshotSize;
+    assert snapshotSize <= mapBuilder.getSize();
+  }
+
+  static EnsoHashMap createWithBuilder(EnsoHashMapBuilder mapBuilder, int snapshotSize) {
+    return new EnsoHashMap(mapBuilder, snapshotSize);
+  }
+
+  static EnsoHashMap createEmpty(HashCodeAnyNode hashCodeAnyNode, EqualsAnyNode equalsNode) {
+    return new EnsoHashMap(EnsoHashMapBuilder.create(hashCodeAnyNode, equalsNode), 0);
+  }
+
+  EnsoHashMapBuilder getMapBuilder() {
+    return mapBuilder;
+  }
+
+  Object getCachedVectorRepresentation() {
+    assert cachedVectorRepresentation != null;
+    return cachedVectorRepresentation;
+  }
+
+  boolean isVectorRepresentationCached() {
+    return cachedVectorRepresentation != null;
+  }
+
+  @TruffleBoundary
+  void cacheVectorRepresentation() {
+    assert cachedVectorRepresentation == null : "Caching vector repr should be done at most once";
+    Object[] keys = new Object[snapshotSize];
+    Object[] values = new Object[snapshotSize];
+    int arrIdx = 0;
+    for (StorageEntry entry : mapBuilder.getStorage().getValues()) {
+      if (entry.index() <= snapshotSize) {
+        keys[arrIdx] = entry.key();
+        values[arrIdx] = entry.value();
+        arrIdx++;
+      }
+    }
+    cachedVectorRepresentation =
+        Vector.fromArray(HashEntriesVector.createFromKeysAndValues(keys, values));
+  }
+
+  @Builtin.Method
+  @Builtin.Specialize
+  public static EnsoHashMap empty(
+      @Cached HashCodeAnyNode hashCodeNode, @Cached EqualsAnyNode equalsNode) {
+    return createEmpty(hashCodeNode, equalsNode);
+  }
+
+  @ExportMessage
+  boolean hasHashEntries() {
+    return true;
+  }
+
+  @ExportMessage
+  int getHashSize() {
+    return snapshotSize;
+  }
+
+  @ExportMessage
+  boolean isHashEntryExisting(Object key) {
+    return isEntryInThisMap(mapBuilder.get(key));
+  }
+
+  @ExportMessage
+  boolean isHashEntryReadable(Object key) {
+    return isHashEntryExisting(key);
+  }
+
+  @ExportMessage
+  Object readHashValue(Object key) throws UnknownKeyException {
+    StorageEntry entry = mapBuilder.get(key);
+    if (isEntryInThisMap(entry)) {
+      return entry.value();
+    } else {
+      throw UnknownKeyException.create(key);
+    }
+  }
+
+  @ExportMessage
+  Object getHashEntriesIterator(@CachedLibrary(limit = "3") InteropLibrary interop) {
+    if (!isVectorRepresentationCached()) {
+      cacheVectorRepresentation();
+    }
```

Comment:
shouldn't this cache fill logic be inside `getCachedVectorRepresentation`? That way one does not have to very that you checked the cache first on invocation

Reply:
+1 to that, much easier to avoid errors that way